File: 2. Structural Design Patterns/Facade/c++/FacadeTemplate.cpp
#include <iostream>
using namespace std;

// Subsystem A
class SubsystemA {
public:
    void operationA() {
        cout << "SubsystemA: Operation A\n";
    }
};

// Subsystem B
class SubsystemB {
public:
    void operationB() {
        cout << "SubsystemB: Operation B\n";
    }
};

// Subsystem C
class SubsystemC {
public:
    void operationC() {
        cout << "SubsystemC: Operation C\n";
    }
};

// Facade
class Facade {
private:
    SubsystemA a;
    SubsystemB b;
    SubsystemC c;
public:
    void simpleOperation() {
        cout << "Facade: simpleOperation calls:\n";
        a.operationA();
        b.operationB();
        c.operationC();
    }
};

File: 2. Structural Design Patterns/Facade/Python/FacadeTemplate.py
# Facade Pattern – Python Template

class SubsystemA:
    def operation_a(self):
        return "SubsystemA operation"

class SubsystemB:
    def operation_b(self):
        return "SubsystemB operation"

class SubsystemC:
    def operation_c(self):
        return "SubsystemC operation"

class Facade:
    def __init__(self):
        self.subsystem_a = SubsystemA()
        self.subsystem_b = SubsystemB()
        self.subsystem_c = SubsystemC()
    
    def operation(self):
        results = []
        results.append(self.subsystem_a.operation_a())
        results.append(self.subsystem_b.operation_b())
        results.append(self.subsystem_c.operation_c())
        return "\n".join(results)

def main():
    facade = Facade()
    print(facade.operation())

if __name__ == "__main__":
    main()
